Cypher interpolation

jeffmay opened the pull request:
@audax-shreyaspurohit @mingy711 @macmacbr @danstaley @egprentice

Added Cypher String interpolation, CypherValue, and some more unit tests. This should make it easier to use direct Cypher queries.

Commit:
Added CypherValue and Cypher string interpolation

- Compile Statements by embedding Cypher.params() into cypher""
- Rename Statement.statement to Statement.template
- Added AST for cypher values
- Prevent mixed type arrays at compile-time
- Prevent empty arrays at compile-time
- Added ScalaCheck generators and shrinks for CypherValue
- Switched from using JsObject to using CypherProps
- Updated lots of documentation

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala:
```diff
@@ -0,0 +1,30 @@
+
+import scala.language.{dynamics, implicitConversions}
+
+object Cypher {
```
FAIL: Document all of this.

jeffmay replied:
DONE. See below this line.

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherInterpolation.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+class InvalidCypherException(val message: String, val errors: Seq[CypherResultInvalid]) extends Exception({
```
FAIL: Document.

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherValue.scala:
```diff
@@ -0,0 +1,88 @@
+
+import scala.annotation.implicitNotFound
+
+sealed trait CypherValue {
```
FAIL: Document

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/scalacheck/CypherValueGenerators.scala:
```diff
@@ -0,0 +1,182 @@
+//
+//  implicit def arbCypherDoubleArray(implicit arbDouble: Arbitrary[Double]): Arbitrary[CypherArray[CypherDouble]] = Arbitrary {
+//    Gen.listOf(arbDouble.arbitrary.map(CypherDouble(_))).map(CypherArray(_))
+//  }
```
FAIL: Remove

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/results.scala:
```diff
@@ -69,7 +70,7 @@ object StatementResultRow {
 
 /**
-  * Stats about the effects of executing the [[Statement]].
+  * Stats about the effects of executing the [[me.jeffmay.neo4j.client.cypher.Statement]].
```
Use relative import

jeffmay commented on ws/src/main/scala/me/jeffmay/neo4j/client/rest/package.scala:
```diff
@@ -1,8 +1,22 @@
+    case CypherByte(v) => Json.toJson(new String(Array(v)))
+    case CypherChar(v) => Json.toJson(v.toString)
+    case CypherShort(v) => Json.toJson(v)
+  }
```
~~FAIL: This should have a play-json test spec. Maybe make this a Format[CypherValue] to handle round-trip serialization in the future.~~

I'm revoking this fail because we only ever write the server. Once round-trip is supported, I can bring in play-json-tests to support symmetric testing.

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala:
```diff
@@ -0,0 +1,44 @@
+
+  // TODO: Use sourcecode to get the value name this is assigned to
+  //  def params(): Params = new Params("props", Map.empty)
+  def params(namespace: String): Params = new Params(namespace)
```
Document `params`

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/TxnRef.scala:
```diff
@@ -4,7 +4,20 @@ import play.api.libs.json.{JsString, Writes}
+  *       all requests for a given transaction are sent to the same Neo4j instance.
+  *
+  * @param url the URL to commit the transaction
+  * @param host the host address of the Neo4j instance on which the transaction is open
```
doesn't url consist of the host? Cas if it does you should be able to get it without asking someone to set it. But if url is actually uri that you use with host to form url then this should change to uri.

jeffmay replied:
URL does consist of the host. I just wanted to allow access to the original value for logging and compatibility purposes. Splitting it up is just to enable more intelligent handling of transactions (like grouping by host or creating a map with the integer transaction id as a key). Probably not that useful.

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/TxnResponse.scala:
```diff
@@ -1,5 +1,6 @@
 package me.jeffmay.neo4j.client
 
+import me.jeffmay.neo4j.client.cypher.Statement
```
y?

jeffmay replied:
It is linked in the ScalaDocs. I moved the class out of the package.

jtsmith0107 commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherArg.scala:
```diff
@@ -0,0 +1,119 @@
+  * @param result the valid result.
+  */
+case class CypherResultValid[+T <: CypherArg](result: T) extends CypherResult[T] {
+  override def isValid: Boolean = true
```
final val?

jeffmay replied:
I could make this a `final def`. I don't want to allocate space for this value, when it isn't called that often. Also, static dispatch to a `def` with a constant is quick.

jtsmith0107 commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherInterpolation.scala:
```diff
@@ -0,0 +1,57 @@
+      .map { case (namespace, fields) =>
+        val props: CypherProps = fields.groupBy(_.id).map { case (name, values) =>
+          // TODO: Provide better error message by pointing to location in query where the error occurs
+          // Use Li Haoyi's sourcecode macros?
```
I often ask myself if I should use lihaoyi projects

jeffmay replied:
I'm pretty tempted. There are already two use cases that would be served well by it.

jtsmith0107 commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherInterpolation.scala:
```diff
@@ -0,0 +1,57 @@
+        val props: CypherProps = fields.groupBy(_.id).map { case (name, values) =>
+          // TODO: Provide better error message by pointing to location in query where the error occurs
+          // Use Li Haoyi's sourcecode macros?
+          require(values.tail.isEmpty, s"'$name' has already been defined for the props object named '$namespace'")
```
might be nice if users can expect exceptions to be in the CypherException type tree instead of whatever `require` throws

jeffmay replied:
Good idea. I should make an exception for this case. Something like a `DuplicatePropertyNameException`.

jtsmith0107 commented on core/src/test/scala/me/jeffmay/neo4j/client/cypher/CypherValueSpec.scala:
```diff
@@ -0,0 +1,33 @@
+
+    "not allow building mixed arrays" in {
+      CypherArray(Seq(CypherInt(1), CypherInt(2))) // compiles
+      assertDoesNotCompile("CypherArray(Seq(CypherInt(1), CypherBoolean(false)))")
```
different then IllTyped?

jeffmay replied:
ScalaTest's version of it.

jtsmith0107 commented on core/src/test/scala/me/jeffmay/neo4j/client/cypher/CypherValueSpec.scala:
```diff
@@ -0,0 +1,33 @@
+  "CypherArray" should {
+
+    "not allow building mixed arrays" in {
+      CypherArray(Seq(CypherInt(1), CypherInt(2))) // compiles
```
Possibly `CypherArray(Seq(CypherInt(1), CypherInt(2))) =:= CypherArray[CypherInt]`

jtsmith0107 commented on ws/src/main/scala/me/jeffmay/neo4j/client/rest/package.scala:
```diff
@@ -1,8 +1,22 @@
 
-import play.api.libs.json.{Json, Reads}
+import me.jeffmay.neo4j.client.cypher._
+import play.api.libs.json.{JsArray, Json, Reads, Writes}
```
Do you use play json in the base project? perhaps should only be pulled into "test"

jtsmith0107 replied:
Yeah looks like you replaced play-json with your own Cypher Props

jeffmay replied:
This is the `ws` project which uses Play Json and Play WS. I wrote a converter from `CypherValue` to `JsValue`

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherArg.scala:
```diff
@@ -0,0 +1,119 @@
+  * into a Cypher [[Statement]].
+  */
+sealed trait CypherArg {
+  def template: String
```
document what does template refers to here

jeffmay replied:
:+1: Good idea

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala:
```diff
@@ -0,0 +1,62 @@
+  def params(namespace: String): Params = new Params(namespace)
+
+  class Params private[Cypher] (val namespace: String) extends Dynamic {
+    def applyDynamic[T](name: String)(value: T)(implicit writer: CypherWrites[T, CypherValue]): CypherArg = {
```
overrides something in Dynamic?

jeffmay replied:
I learned everything about `Dynamic` here: http://stackoverflow.com/questions/15799811/how-does-type-dynamic-work-and-how-to-use-it

It's really powerful. It mixes type-safe and dynamic language features into one.

Basically, this method is parameterized and returns a type-safe value, requiring implicit evidence for even more type-safety, however, the method name is dynamic. All method calls that are not statically defined fall into the `applyDynamic` method (there are other ones you can choose to implement as well). `applyDynamic` is treated like a method call where the name is a string and the value is the argument passed to the method with that name.

jeffmay replied:
Short answer... not exactly. It can't utilize the `override` keyword, because the type signature isn't defined until this class.

audax-shreyaspurohit replied:
interesting. I think groovy used to have similar concept on adding a method dynamically on a class.

audax-shreyaspurohit replied:
yea, cant override as its a marker interface that is filled by compiler later.

jeffmay commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala:
```diff
@@ -0,0 +1,62 @@
+    *
+    * @param fields the key-value pairs of the properties and their associated [[CypherValue]].
+    *
+    * @return
```
FAIL: document how later values override earlier values.

Commit:
Updated REST exceptions and added common CypherException root.

- Moved and deprecated old location / names for REST exceptions
- Left deprecated type aliases for good measure
- Updated the documentation

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/Cypher.scala:
```diff
@@ -0,0 +1,62 @@
+    *   val id: String = "1"
+    *   val props = Cypher.params("props")
+    *   val stmt = cypher"MATCH (n { id: ${props.id(id)} }) RETURN n"
+    *   stmt.template // "MATCH (n { id: {props}.id }) RETURN n"
```
shouldn't this be `"MATCH (n { id: {id} }) RETURN n"` if the stmt.parameters is Map("id" -> "1") ?

jeffmay replied:
Checkout http://neo4j.com/docs/stable/rest-api-transactional.html for how the the template works.

audax-shreyaspurohit commented on core/src/main/scala/me/jeffmay/neo4j/client/cypher/CypherWritesPrimitive.scala:
```diff
@@ -0,0 +1,80 @@
+package me.jeffmay.neo4j.client.cypher
+
+trait CypherWrites[-T, +V <: CypherValue] {
```
document this trait please. Its modeled based on Json writes I am assuming.

jeffmay replied:
:+1: to documenting

audax-shreyaspurohit commented on ws/src/main/scala/me/jeffmay/neo4j/client/rest/package.scala:
```diff
@@ -1,8 +1,22 @@
+    case CypherBoolean(v) => Json.toJson(v)
+    case CypherLong(v) => Json.toJson(v)
+    case CypherFloat(v) => Json.toJson(v)
+    case CypherByte(v) => Json.toJson(new String(Array(v)))
```
UTF-8?

jeffmay replied:
Good call.

audax-shreyaspurohit commented:
lgtm

macmacbr commented:
LGTM
